drivers/esp-hosted/esp_hosted_hal.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2023 Arduino SA
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * ESP-Hosted WiFi HAL.
+ */
+
+#include "py/mphal.h"
+
+#if MICROPY_PY_NETWORK_ESP_HOSTED
+
+#include <stdint.h>
+#include <string.h>
+
+#include "py/runtime.h"
+#include "modmachine.h"
+#include "extmod/machine_spi.h"
+#include "mpconfigboard.h"
+
+#include "esp_hosted_hal.h"
+#include "esp_hosted_wifi.h"
+
+MP_WEAK int esp_hosted_hal_init(uint32_t mode) {
+    if (mode == ESP_HOSTED_MODE_BT) {
+        // Remove all network interfaces and reset wifi state.
+        esp_hosted_wifi_deinit();
+    }
+
+    // Perform a hard reset and set pins to their defaults.
+    esp_hosted_hal_deinit();
+
+    if (mode == ESP_HOSTED_MODE_BT) {
+        // For Bluetooth mode, init is done.
+        return 0;
+    }
+
+    mp_hal_pin_input(MICROPY_HW_WIFI_HANDSHAKE);
+    mp_hal_pin_input(MICROPY_HW_WIFI_DATAREADY);
+
+    // Initialize SPI.
+    mp_obj_t args[] = {
+        MP_OBJ_NEW_SMALL_INT(MICROPY_HW_WIFI_SPI_ID),
+        MP_OBJ_NEW_SMALL_INT(MICROPY_HW_WIFI_SPI_BAUDRATE),
+        MP_OBJ_NEW_QSTR(MP_QSTR_polarity), MP_OBJ_NEW_SMALL_INT(1),
+    };
+
+    MP_STATE_PORT(mp_wifi_spi) =
+        MP_OBJ_TYPE_GET_SLOT(&machine_spi_type, make_new)((mp_obj_t)&machine_spi_type, 2, 1, args);
+
+    // SPI might change the direction/mode of CS pin,
+    // set it to GPIO again just in case.
+    mp_hal_pin_output(MICROPY_HW_WIFI_SPI_CS);
+    mp_hal_pin_write(MICROPY_HW_WIFI_SPI_CS, 1);
+    return 0;
+}
+
+MP_WEAK int esp_hosted_hal_deinit(void) {
+    mp_hal_pin_output(MICROPY_HW_ESP_HOSTED_GPIO0);
+    mp_hal_pin_output(MICROPY_HW_ESP_HOSTED_RESET);
+
+    mp_hal_pin_output(MICROPY_HW_WIFI_SPI_CS);
+    mp_hal_pin_write(MICROPY_HW_WIFI_SPI_CS, 1);
```

Comment:
esp_hosted_hal_deinit() is called after the BLE interface is set up in mp_bluetooth_hci_uart_init(). Re-configuring the CS pin may override a configuration made for BLE UART. Are you sure that it is strictly needed here?

Reply:
The SPI CS pin should be a different pin, so normally this is not a problem, it's meant to leave the module in a known state.

Reply:
It is possible that a UART suitable for BLE is at the same pins as the SPI pin. Examples: MIMXRT, SAMD. And the default setting of the esp_hosted firmware used the same pins for WiFi and BLE, maybe assuming to use UART over SPI as communication method. I do not know, if the esp_hosted firmware can serve BLE and WiFi at the same time. If yes, the  a UART and SPI interface must be at separate pins, like it is configured in this PR, and in my trial port for MIMXRT.
If a board at a given time uses neither BLE of WiFI, the state of the CS pin is undefined. If the state of CS matters, both BLE and WiFi code paths shall set CS in their init sequence.

Reply:
It might be possible to remove it, but it will require retesting WiFi, Bluetooth and switching from WiFi to Bluetooth and vice versa.